src/evaluators/source-4-3.js

```diff
@@ -0,0 +1,808 @@
+    } else if (is_compound_function(fun)) {
+        const body = function_body(fun);
+        const locals = function_locals(fun);
+        const names = insert_all(map(name_of_name, function_parameters(fun)), locals);
+        const temp_values = map(x => no_value_yet, locals);
+        const values = append(args, temp_values);
+
+        function_body(fun)(
+            extend_environment(
+                names,
+                values,
+                function_environment(fun)),
+            succeed,
+            fail);
+    } else {
+        error(fun, "unknown function type in " +
+            "execute_application");
+    }
+}
+
+function analyze_amb(exp) {
+    const cfuns = map(analyze, amb_choices(exp));
+    return (env, succeed, fail) => {
+        function try_next(choices) {
+            return is_null(choices) ?
+                fail() :
+                head(choices)(env,
+                    succeed,
+                    () =>
+                        try_next(tail(choices)));
+        }
+        return try_next(cfuns);
+    };
+}
+
+const input_prompt = "// Amb-Eval input:";
+const output_prompt = "// Amb-Eval value:";
+
+function announce_output(s) {
+    display(s);
+}
+
+function driver_loop() {
+    function internal_loop(try_again) {
+        const input = prompt(input_prompt);
+        if (input === "try-again") {
+            try_again();
+        } else {
+            display("// Starting a new problem ");
+            ambeval(input,
+                the_global_environment,
+                // ambeval success
+                (val, next_alternative) => {
+                    announce_output(output_prompt);
+                    user_print(val);
+                    return internal_loop(next_alternative);
+                },
+                // ambeval failure
+                () => {
+                    announce_output(
+                        "// There are no more values of");
+                    user_print(input);
+                    return driver_loop();
+                });
+        }
+    }
+    return internal_loop(
+        () => {
+            display("// There is no current problem");
+            return driver_loop();
+        });
+}
+
+function user_print(object) {
+    return is_compound_function(object) ?
+        "function" +
+        stringify(function_parameters(object)) +
+        stringify(function_body(object)) +
+        "<environment>" :
+        object;
+}
+
+let try_again = null;
+
+function parse_and_run(str) {
+    const exprs = make_block(parse(str));
+    ambeval(exprs,
+        the_global_environment,
+        // ambeval success
+        (val, next_alternative) => {
+            announce_output(output_prompt);
+            display(user_print(val));
+            try_again = next_alternative;
+        },
+        // ambeval failure
+        () => {
+            announce_output(
+                "// There are no more values of");
+            display(user_print(str));
+        });
+}
```

Comment:
Examples missing:

List the examples from the textbook and make sure they all work

Reply:
Added example in a15a077